### PR TITLE
update SQLite, create Nuget, add decoding

### PR DESCRIPTION
**Pull request by ptahmose**
## Description

This PR includes the following changes:
* update SQLite to 3.45.2
* add functionality to decode compressed data to imgdoc2API and imgdoc2Net
* add functionality to create a Nuget package (note: this Nuget is currently **not** yet being published)

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

* locally
* unit-tests

## Checklist:

- [x] I followed the Contributing Guidelines.
- [x] I did a self-review.
- [ ] I commented my code, particularly in hard-to-understand areas.
- [ ] I updated the documentation.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

### CODE REVIEW

**Review by DaveyJonesBitPail — DISMISSED** [review]
Looks good, might be a good idea for future effort to have imgdoc2 consume sqlite via vcpkg in order to simplify upgrade step

File: misc/test_convert.ps1 [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# SPDX-FileCopyrightText: 2024 Carl Zeiss Microscopy GmbH
+#
+# SPDX-License-Identifier: MIT
+
+# Define the path to the executable
+$executablePath = "D:\dev\MyGitHub\ptahmose_imgdoc2\out\build\x64-Debug\convczi\convczi.exe"
```
Comment (DaveyJonesBitPail) [inline misc/test_convert.ps1:6]: Not a huge problem, but may want to remove hardcoded paths here and use arguments or values set by a .env file for transient, user-specific, environment variables ([DotEnv](https://github.com/stopthatastronaut/poshdotenv) for example is a module that implements the logic for this)

File: libimgdoc2/src/db/IDbStatement.h
Comment (DaveyJonesBitPail) [inline libimgdoc2/src/db/IDbStatement.h:1]: Small note the comments on line 31-33 are one space behind where they ought to be

File: libimgdoc2/inc/IDocQuery3d.h [hunk 2]
```diff
@@ -25,35 +25,35 @@ namespace imgdoc2
...
         /// coordinates, the second by other "per tile data". The functor is called for each tile which matches the query. If the functor
-        /// returns false, the enumeration is canceled, and no more calls to the functor will occur any more. The two query clauses are
+        /// returns false, the enumeration is canceled, and no more calls to the functor will occur anymore. The two query clauses are
         /// logically ANDed together.
         /// \param clause        The query clause (dealing with dimension indexes).
         /// \param tileInfoQuery The query clause (dealing with other "per tile data").
-        /// \param func          A functor which we will called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
-        ///                      more calls to the functor will occur any more.
+        /// \param func          A functor which we will be called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
```
Comment (DaveyJonesBitPail) [inline libimgdoc2/inc/IDocQuery3d.h:39]:
```suggestion
        /// \param func          A functor which will be called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
```

File: libimgdoc2/inc/IDocQuery3d.h [hunk 3]
```diff
@@ -25,35 +25,35 @@ namespace imgdoc2
...
+        ///                      more calls to the functor will occur anymore.
         virtual void Query(const imgdoc2::IDimCoordinateQueryClause* clause, const imgdoc2::ITileInfoQueryClause* tileInfoQuery, const std::function<bool(imgdoc2::dbIndex)>& func) = 0;
 
         /// Gets tiles intersecting the specified cuboid (and satisfying the other criteria).
         /// \param  cuboid            The cuboid.
         /// \param  coordinate_clause The coordinate clause.
         /// \param  tileinfo_clause   The tileinfo clause.
-        /// \param  func              A functor which we will called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
-        ///                           more calls to the functor will occur any more.
+        /// \param  func              A functor which we will be called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
```
Comment (DaveyJonesBitPail) [inline libimgdoc2/inc/IDocQuery3d.h:47]:
```suggestion
        /// \param  func              A functor which will be called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
```

File: libimgdoc2/inc/IDocQuery3d.h [hunk 4]
```diff
@@ -25,35 +25,35 @@ namespace imgdoc2
...
+        ///                           more calls to the functor will occur anymore.
         virtual void GetTilesIntersectingCuboid(const imgdoc2::CuboidD& cuboid, const imgdoc2::IDimCoordinateQueryClause* coordinate_clause, const imgdoc2::ITileInfoQueryClause* tileinfo_clause, const std::function<bool(imgdoc2::dbIndex)>& func) = 0;
 
         /// Gets tiles intersecting with the specified plane (and satisfying the other criteria).
         ///
         /// \param   plane               The plane.
         /// \param   coordinate_clause   The coordinate clause.
         /// \param   tileinfo_clause     The tileinfo clause.
-        /// \param   func                A functor which we will called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
+        /// \param   func                A functor which we will be called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
```
Comment (DaveyJonesBitPail) [inline libimgdoc2/inc/IDocQuery3d.h:56]:
```suggestion
        /// \param   func                A functor which will be called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
```

File: libimgdoc2/inc/IDocQuery3d.h [hunk 5]
```diff
@@ -25,35 +25,35 @@ namespace imgdoc2
...
+        /// \param [out]    coordinate      If non-null and the operation is successful, the tile-coordinate will be put here.
         /// \param [out]    info            If non-null and the operation is successful, the logical position will be put here.
         /// \param [out]    brick_blob_info If non-null and the operation is successful, the brick-blob-info will be put here.
         virtual void ReadBrickInfo(imgdoc2::dbIndex idx, imgdoc2::ITileCoordinateMutate* coordinate, imgdoc2::LogicalPositionInfo3D* info, imgdoc2::BrickBlobInfo* brick_blob_info) = 0;
 
         /// Query the tiles table. The two query clauses are used to filter the tiles. The first clause is used to filter the tiles by their
         /// coordinates, the second by other "per tile data". The functor is called for each tile which matches the query. If the functor
-        /// returns false, the enumeration is canceled, and no more calls to the functor will occur any more. The two query clauses are
+        /// returns false, the enumeration is canceled, and no more calls to the functor will occur anymore. The two query clauses are
```
Comment (DaveyJonesBitPail) [inline libimgdoc2/inc/IDocQuery3d.h:35]: anymore can be dropped here without losing any meaning in the sentence. Sounds better to me but only opinion based

File: libimgdoc2/inc/IDocQuery2d.h [hunk 6]
```diff
@@ -33,12 +33,12 @@ namespace imgdoc2
...
        /// coordinates, the second by other "per tile data". The functor is called for each tile which matches the query. If the functor
-       /// returns false, the enumeration is canceled, and no more calls to the functor will occur any more. The two query clauses are
+       /// returns false, the enumeration is canceled, and no more calls to the functor will occur anymore. The two query clauses are
        /// logically ANDed together.
        /// \param clause        The query clause (dealing with dimension indexes).
        /// \param tileInfoQuery The query clause (dealing with other "per tile data").
-       /// \param func          A functor which we will called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
-       ///                      more calls to the functor will occur any more.
+       /// \param func          A functor which we will be called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
```
Comment (DaveyJonesBitPail) [inline libimgdoc2/inc/IDocQuery2d.h:40]:
```suggestion
       /// \param func          A functor which will be called, passing in the index of tiles matching the query. If the functor returns false, the enumeration is canceled, and no
```

File: libimgdoc2/inc/IDocQuery2d.h [hunk 7]
```diff
@@ -33,12 +33,12 @@ namespace imgdoc2
 
        /// Query the tiles table. The two query clauses are used to filter the tiles. The first clause is used to filter the tiles by their
        /// coordinates, the second by other "per tile data". The functor is called for each tile which matches the query. If the functor
-       /// returns false, the enumeration is canceled, and no more calls to the functor will occur any more. The two query clauses are
+       /// returns false, the enumeration is canceled, and no more calls to the functor will occur anymore. The two query clauses are
```
Comment (DaveyJonesBitPail) [inline libimgdoc2/inc/IDocQuery2d.h:36]: anymore could also be dropped (opinion)

File: libimgdoc2/inc/ClassFactory.h [hunk 8]
```diff
@@ -18,6 +19,11 @@ namespace imgdoc2
     class ClassFactory
     {
     public:
+        /// Gets built-time information about the libimgdoc2 library.
```
Comment (DaveyJonesBitPail) [inline libimgdoc2/inc/ClassFactory.h:22]:
```suggestion
        /// Gets build-time information about the libimgdoc2 library.
```

File: libimgdoc2/CMakeLists.txt [hunk 9]
```diff
@@ -104,4 +154,7 @@ target_compile_definitions(libimgdoc2 PUBLIC _SQLITEIMGDOCSTATICLIB=1)
 
 target_include_directories(libimgdoc2 INTERFACE
   $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/inc> 
-  )
\ No newline at end of file
+  )
+
+  # add the binary tree to the search path for include files so that we will find libimgdoc2_config.h
```
Comment (DaveyJonesBitPail) [inline libimgdoc2/CMakeLists.txt:159]: leading whitespace difference in comment and line it refers to

File: imgdoc2API/imgdoc2API.h [hunk 10]
```diff
@@ -63,6 +65,19 @@ typedef bool(LIBIMGDOC2_STDCALL* MemTransferReserveFunctionPointer)(std::intptr_
...
+typedef bool(LIBIMGDOC2_STDCALL* AllocMemoryFunctionPointer)(std::uint64_t size, AllocationObject* allocation_object); // NOLINT(readability/casting)
+
+/// Gets version information. This function is used to retrieve the version information of the library.
+/// Note that the specified allocate_memory_function is used to allocate memory for the strings in the VersionInfoInterop structure.
+/// If the allocate_memory_function is null, then the strings will not be returned.
+///
+/// \param [out]    version_info                Structure where the version information will be placed.
+/// \param          allocate_memory_function    If non-null, this "allocate memory function" will be used to allocate memory for the strings in the VersionInfoInterop structure
+///  
+/// \returns An error-code indicating success or failure of the operation.                                                                                       to be returned.
```
Comment (DaveyJonesBitPail) [inline imgdoc2API/imgdoc2API.h:78]: Unclear of the purpose of the to be returned on the far right here.

File: imgdoc2API/codecsAPI.h [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+#pragma once
+
+#include "imgdoc2API.h"
+#include "importexport.h"
+#include "errorcodes.h"
+#include "bitmapinfointerop.h"
+#include "decodedimageresultinterop.h"
+
+/// Decodes the specific compressed data into an uncompressed bitmap. This destination bitmap is allocated
```
Comment (DaveyJonesBitPail) [inline imgdoc2API/codecsAPI.h:13]: (Opinion) Could be "Decodes the specified". Sounds more natural to me

File: dotnet/imgdoc2cmd/CommandQueryForTilesAndSaveTiles.cs [hunk 12]
```diff
@@ -33,68 +28,52 @@ protected override void ProcessTiles(List<long> tiles, Options options)
...
-                        return bitmapSource;
+                        return CreateImageImageSharp(tileBlobInfo.Tile2dBaseInfo.PixelType, tileData, tileBlobInfo.Tile2dBaseInfo.PixelWidth, tileBlobInfo.Tile2dBaseInfo.PixelHeight);
                     }
                 default:
                     throw new InvalidOperationException("Could not process tile.");
             }
         }
 
-        private static PixelFormat ImgDoc2PixelTypeToWpfPixelFormats(PixelType pixelType)
+        private static Image CreateImageImageSharp(PixelType pixelType, Memory<byte> data,int width,int height)
```
Comment (DaveyJonesBitPail) [inline dotnet/imgdoc2cmd/CommandQueryForTilesAndSaveTiles.cs:64]:
```suggestion
        private static Image CreateImageImageSharp(PixelType pixelType, Memory<byte> data, int width, int height)
```

**Review by DaveyJonesBitPail — APPROVED** [review]
Approved with minor suggestions

**Review by ReneBoeckelmann — APPROVED** [review]

**Review by ReneBoeckelmann — APPROVED** [review]

File: imgdoc2API/codecsAPI.h (hunk 11 above)
Comment (ptahmose) [reply imgdoc2API/codecsAPI.h:13]: done, thanks

File: imgdoc2API/imgdoc2API.h (hunk 10 above)
Comment (ptahmose) [reply imgdoc2API/imgdoc2API.h]: No purpose I can think of. Fixed, thanks.

File: libimgdoc2/CMakeLists.txt (hunk 9 above)
Comment (ptahmose) [reply libimgdoc2/CMakeLists.txt]: fixed, thanks

File: libimgdoc2/src/db/IDbStatement.h
Comment (ptahmose) [reply libimgdoc2/src/db/IDbStatement.h:1]: fixed, thanks

File: misc/test_convert.ps1 (hunk 1 above)
Comment (ptahmose) [reply misc/test_convert.ps1:9]: I added a comment stating that this is work-in-progress and not meant to be used as-is.

File: libimgdoc2/inc/IDocQuery2d.h (hunk 7 above)
Comment (ptahmose) [reply libimgdoc2/inc/IDocQuery2d.h]: done, thanks!

File: libimgdoc2/inc/IDocQuery2d.h (hunk 6 above)
Comment (ptahmose) [reply libimgdoc2/inc/IDocQuery2d.h:40]: done, thanks!

File: libimgdoc2/inc/IDocQuery3d.h (hunk 5 above)
Comment (ptahmose) [reply libimgdoc2/inc/IDocQuery3d.h:35]: done, thanks!

**Review by ReneBoeckelmann — APPROVED** [review]

**Pull request merged**